### PR TITLE
[auto-toolstate] Upload the toolstate result to an external git repository, and removes BuildExpectation

### LINKED ISSUE #45861: Automate management of broken tools
Rather than using the tools manifest for opting-in to breaking a tool.

The rough plan: on master (but not beta or stable) the CI would ignore failing to build or test some tools (RLS, Clippy, etc). However, some bot would note when a tool started to fail and email interested parties (and/or display the info in a dashboard).

This should lower the friction for contributors working in the Rust repo when they make tool-breaking changes. However, we might miss some 'true positive' failure in the tools (e.g., #45820).

### PR DESCRIPTION
**Pull request by kennytm**
This PR consists of 3 commits.

1. (Steps 4–6) The `toolstate.json` output previously collected is now pushed to the https://github.com/rust-lang-nursery/rust-toolstate repository. 
2. (Step 7) Revert commit ab018c7, thus removing all traces of `BuildExpectation` and `toolstate.toml`.
3. (Step 8) Adjust CONTRIBUTION.md for the new procedure.

These are the last steps of #45861. After this PR, the toolstate will be automatically computed and published to https://rust-lang-nursery.github.io/rust-toolstate/. There is no need to manage toolstate.toml again.

Closes #45861.

### CODE REVIEW

**Comment by kennytm** [comment]
WIP —

- Re-verify toolstate is correctly calculated and uploaded:

    - [ ] 1st try — obtain current status (miri: broken, everything else: testing)
    - [ ] 2nd try — deliberately break the testing code of several tools to see the effect 
    - [ ] 3rd try — restore testability of these tools

- [ ] Change `publish.py` to actually mention people 

- [ ] Administrative — move kennytm/rust-toolstate to rust-lang-nursery/rust-toolstate 

- [ ] Blocked by #45002, which reenables miri

**Comment by kennytm** [comment]
clippy is broken? 🤔

**Comment by oli-obk** [comment]
Jup, I'm fixing it :)

**Comment by oli-obk** [comment]
You can update the submodule to clippy's master branch. It was just some whitespace changes in ui output

**Comment by kennytm** [comment]
@oli-obk OK thanks 😄 It is worrying that the CI did not catch this, so I've updated `checktools.sh` to stop relying on `$?`, instead `grep` the test result directly.

**Comment by kennytm** [comment]
@bors try 

Second try:

* Updated `clippy` to master to unbreak it
* Overwrite a test in `rustfmt` to try to make the state fallback to Compiling.

**Comment by kennytm** [comment]
@oli-obk Just to let you know, clippy's dogfood test [failed](https://travis-ci.org/rust-lang/rust/jobs/312935100) to run due to read-only file system. 

```
[01:07:12] running 2 tests
[01:07:12] test compile_test ... FAILED
[01:07:12] test dogfood ... FAILED
[01:07:12] 
[01:07:12] failures:
[01:07:12] 
[01:07:12] ---- compile_test stdout ----
[01:07:12] 	thread 'compile_test' panicked at 'called `Result::unwrap()` on an `Err` value: Error { repr: Os { code: 30, message: "Read-only file system" } }', /checkout/src/libcore/result.rs:906:4
[01:07:12] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:07:12] 
[01:07:12] ---- dogfood stdout ----
[01:07:12] 	thread 'dogfood' panicked at 'called `Result::unwrap()` on an `Err` value: Error { repr: Os { code: 2, message: "No such file or directory" } }', /checkout/src/libcore/result.rs:906:4
[01:07:12] 
[01:07:12] 
[01:07:12] failures:
[01:07:12]     compile_test
[01:07:12]     dogfood
[01:07:12] 
[01:07:12] test result: FAILED. 0 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out
[01:07:12] 
[01:07:12] error: test failed, to rerun pass '--test compile-test'
```

Looks like `clippy` was not tested at all previously 😂, thus there were no report of such failure. https://github.com/rust-lang/rust/blob/f4f18586e755f239b5be3a60a6e3bfd382524210/src/bootstrap/mk/Makefile.in#L53-L65

**Comment by oli-obk** [comment]
ugh... I can't run ui-tests? Is there any folder I'm allowed to write to? I can just dump the ui files there. Maybe `env::var("OUT_DIR")`?

**Comment by kennytm** [comment]
@bors try 

Homu-try is not compatible with Travis's build stages 😓

**Comment by nikomatsakis** [comment]
I don't think I'm a good reviewer for this. @alexcrichton or @Mark-Simulacrum -- perhaps one of you? I'm going to nominate @alexcrichton to start and they can pass it on to whomever is most appropriate.

r? @alexcrichton

**Comment by kennytm** [comment]
@bors try 

Should be the final try.

**Comment by kennytm** [comment]
Now blocked by solson/miri#358.

**Comment by kennytm** [comment]
@bors try 

One final-final try to check the status works after transferring the GitHub repository (likely everything is broken. doesn't matter as long as a commit is pushed.)

**Comment by kennytm** [comment]
Ok ready for review now.

@alexcrichton (cc @oli-obk @RalfJung @eddyb (miri) @Manishearth @llogiq @mcarton @oli-obk (clippy) @nrc (rls+rustfmt)):

About the procedure:

* Tools are allowed to break on nightly channels. However they must all be "Testing" in beta/stable channels. Because we can't just break a tool on beta/stable now, please make sure all tools are fixed before the next beta is branched off (before Jan 5th or Feb 16th, depending on timing).

* The `toolstate.toml` file is deleted. Whenever a tool is broken or fixed on nightly, commits like rust-lang-nursery/rust-toolstate@42f550bcfe631069a865af5ba1f45b6eaf4f5315 and rust-lang-nursery/rust-toolstate@74a160cc8ac9d6a2d714d142fbcbcac45aaefff6 will be issued *after* a PR is merged. (The `%` in the commit messages will be replaced by `@` so the commit will actually appear in GitHub notifications.)

* Tool maintainer details are documented in https://github.com/rust-lang-nursery/rust-toolstate.

* Contributor details in CONTRIBUTING.md are updated.

About the code in this PR:

* As stated in https://github.com/rust-lang/rust/pull/46554#issuecomment-349928650, it seems relying on exit code is too fragile to catch build failure. The tools job in beta/stable is now considered "failing" whenever the produced toolstate contains the word "Broken" or "Compiling".

* The commit message format is controlled by `src/ci/docker/x86_64-gnu-tools/publish.py`. (This one cannot be rewritten to shell due to JSON parsing. Rewriting to Rust is OK but feels too heavy.)

* The git pull/push actions are controlled by `src/ci/docker/x86_64-gnu-tools/repo.sh`. Currently the commits are pushed by @kennytm-githubbot.

**Comment by alexcrichton** [comment]
This is looking fantastic to me, thanks so much @kennytm!

I see that the tokens are already encrypted here, and while I think you've got the necessary rights to do the Travis one, are you sure the AppVeyor one will work? If not I can get you the text to insert pretty easily I believe.

Code-wise I don't have much to comment on, this all looks great. I'd at least be excited to try it out and see how it goes :)

**Comment by kennytm** [comment]
@alexcrichton Thanks :). My step to generate it is documented in

https://github.com/kennytm/rust/blob/aa9125fcb345cba9ed417c3be5f8fe3653687f6b/src/ci/docker/x86_64-gnu-tools/repo.sh#L43-L52 

which I'm pretty sure will encrypt the token using rust-lang/rust's public key. If the token is wrong GitHub will refuse the push, and we could change the value then.

**Comment by Manishearth** [comment]
> Tools are allowed to break on nightly channels. However they must all be "Testing" in beta/stable channels. Because we can't just break a tool on beta/stable now, please make sure all tools are fixed before the next beta is branched off (before Jan 5th or Feb 16th, depending on timing).


Can do for clippy; though I'm a bit unclear -- does this mean that from now on when clippy is not broken it will be available via rustup? That's pretty neat. By Jan 5th we could probably do some refactorings to make it work better with rustup as an installation method (though I'm inclined to let one cycle pass so that we can tell folks to move to rustup first; it's still distributable in its current form, just that there's a lot of now-unnecessary cruft that can be removed)

File: src/ci/docker/x86_64-gnu-tools/publish.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+import datetime
+import collections
+
+# List of people to ping when the status of a tool changed.
+MAINTAINERS = {
+    'miri': '@oli-obk @RalfJung @eddyb',
+    'clippy-driver': '@Manishearth @llogiq @mcarton @oli-obk',
+    'rls': '@nrc',
+    'rustfmt': '@nrc',
+}
```
Comment (RalfJung) [inline src/ci/docker/x86_64-gnu-tools/publish.py]: Having this list somewhere in a script deep in the `ci/docker` folder may not be the best place. Is there a precedent for where to put such "repository configuration files"?

**Comment by RalfJung** [comment]
I find the output at <https://rust-lang-nursery.github.io/rust-toolstate/> a little confusing -- "Testing" and "Compiling" sound like this is something happening *right now*, i.e., a test / a compile job is in progress currently. I am not sure how to word this better though... "Compilation broken", "Tests fail" and "Good", maybe?

File: src/ci/docker/x86_64-gnu-tools/publish.py (hunk 1 above)
Comment (kennytm) [reply src/ci/docker/x86_64-gnu-tools/publish.py]: @RalfJung This script can be placed elsewhere. Maybe `src/tools/publish_toolstate.py`?

**Comment by kennytm** [comment]
@RalfJung I've reenabled issues on that repo, but let's keep the comment in this PR before it is deployed.

The Broken/Compiling/Testing just mirrors the existing wordings from `toolstates.toml`. Would borrowing the terms from crater clearer?

* build-fail
* test-fail
* test-pass
* unknown

File: src/ci/docker/x86_64-gnu-tools/publish.py (hunk 1 above)
Comment (RalfJung) [reply src/ci/docker/x86_64-gnu-tools/publish.py]: That would certainly help. I was originally thinking of separating configuration and code, but given the size of the script, maybe that's overdoing it.

**Comment by RalfJung** [comment]
> The Broken/Compiling/Testing just mirrors the existing wordings from toolstates.toml.

Fair enough. :)  Now might be a good time to improve this.

> Would borrowing the terms from crater clearer?

Yes, that would be much clearer than the current state.

**Comment by alexcrichton** [comment]
@kennytm ok we can always see how it goes and correct course if need be! (I didn't realize others could log in as the rust-lang organization...)

In any case r=me when you feel this is ready to go, aka when @RalfJung's concern is addressed.

**Comment by kennytm** [comment]
Failure is legit, blocked by #46630

**Comment by oli-obk** [comment]
wooohoo! Thanks for the awesome work @kennytm

**Pull request merged**